### ISSUE TITLE
Support dynamic mixin invocation when transforming mixins

### ISSUE THREAD

**Issue by schatekar** [HenrikJoreteg/templatizer#88]
I am rendering a mixin by specifying it's name dynamically as below

```jade
 +#{"title_" + cv.title.style}(cv.title)
```

In the above code

1. `cv` is the locals object passed into the template
2. Name of the mixin is formed by concatenating the words  `title` and the string value in `cv.title.style`

Following exception is thrown while parsing the above line

```shell
\node_modules\templatizer\lib\transformMixins.js:68
ression.callee.property.value || node.expression.callee.object.property.value;
                                                                                                                ^
TypeError: Cannot read property 'value' of undefined
    at transformMixinCalls (\node_modules\templatizer\lib\transformMixins.js:68:100)
```

The json for `node.expression.callee.object.property` is as below

```js
{ type: 'Identifier',                   
  name: 'jade_mixins',                  
  range: [ 322, 333 ],                  
  parent:                               
   { type: 'MemberExpression',          
     computed: true,                    
     object: [Circular],                
     property:                          
      { type: 'BinaryExpression',       
        operator: '+',                  
        left: [Object],                 
        right: [Object],                
        range: [Object],                
        parent: [Circular],             
        source: [Function],             
        update: [Function: update] },   
     range: [ 322, 360 ],               
     parent:                            
      { type: 'CallExpression',         
        callee: [Circular],             
        arguments: [Object],            
        range: [Object],                
        parent: [Object],               
        source: [Function],             
        update: [Function: update] },   
     source: [Function],                
     update: [Function: update] },      
  source: [Function],                   
  update: [Function: update] }          
```

**Comment by lukekarrys** [comment]
Try setting the `dontRemoveMixins` to `true`.

**Comment by lukekarrys** [comment]
Curious if the above works. If it does then we should probably detect dynamic mixins and disable this functionality for them since it wouldn't even be possible to pull out a mixin from a parent function if it had a dynamic name.

**Comment by lukekarrys** [comment]
You might need to set `dontTransformMixins` to `true` also.

**Comment by schatekar** [comment]
@lukekarrys I tried setting `dontRemoveMixins` to `true` and that is when exception was thrown. 

What does `dontTransformMixins` do?

**Comment by lukekarrys** [comment]
I think `dontTransformMixins: true` is what you'll need until this bug gets fixed. `dontTransformMixins` turns off the functionality described in the [`Mixin Support`](https://github.com/HenrikJoreteg/templatizer#mixin-support) section of the readme. In short, it does some ast parsing so that templates like this


```jade
// - user.jade

mixin user_a(user)
  a #{user.name}

ul
  each user in users
    mixin user_li(user)
```

Get compiled to
```js
exports.user = function () {}

// separate function for the mixin
exports.user.user_a = function () {}
```

**Comment by schatekar** [comment]
@lukekarrys Is your fix going to support dynamic mixin invocation. Because your comment above says that you would rather detect dynamic mixins and throw exception. If that is the case, I would never be able to transform dynamic mixing invocations.

**Comment by lukekarrys** [comment]
No, the fix would be when it encounters dynamic mixins not to pull those out of the parent template file as separate named templates. Because they are dynamic it wouldn't be possible to do what happened in the above example and pull out the `user_a` mixin on to `exports.user` since they would not have a name.

**Comment by lukekarrys** [comment]
Sorry, it looks like I'm confusing dynamic mixin invocation with declaration (the syntax confused me at first since I've never used them before).

So it should be possible to transform the invocation the same as we've already been doing.

Not sure when I'll get to this though, so the only solutions for now would be to not use dynamic mixins or `dontTransformMixins: true`.

**Comment by lukekarrys** [comment]
@schatekar Turned out to be not too difficult. Just published the fix as `v1.5.3`

**Comment by schatekar** [comment]
@lukekarrys Thanks. The template generation works but I am not be able to test the templates just yet. But we should be able to close this ticket on that basis I believe.
